### ISSUE TITLE
OpenGraph images are not statically generated for dynamic routes

### ISSUE THREAD

**Issue by Cretezy** [vercel/next.js#51147]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #1 ZEN SMP PREEMPT_DYNAMIC Sun, 04 Jun 2023 11:52:10 +0000
    Binaries:
      Node: 18.16.0
      npm: 9.5.1
      Yarn: 1.22.10
      pnpm: 8.6.1
    Relevant packages:
      next: 13.4.5-canary.12
      eslint-config-next: 13.4.5
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Metadata (metadata, generateMetadata, next/head, head.js), Static HTML Export (output: "export")

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/reverent-feather-2sycr6

### To Reproduce

- Set `output: "export"` in next.config.js
- Create dynamic path (a path with param, e.g. `app/[id]/page.tsx`) which has `generateStaticParams`
- Create `opengraph-image.tsx` for that path
- Don't set `runtime` (which defaults to `'auto'`/`'nodejs'`)
- Build Next.js (`npm run build`)

### Describe the Bug

The `opengraph-image` is not generated for dynamic paths when statically rendering.

In the output, it's marked as `λ  (Server)  server-side renders at runtime` for the dynamic paths (e.g. `/[id]/opengraph-image`), without listing each generated path (which `/[id]` would have, e.g. `/1`, `/2`,).

I can confirm it generates a `.html` for each path in `out`, but no `opengraph-image` is preset.

(It does work if it has no parameter in the path)

Example build output:
![image](https://github.com/vercel/next.js/assets/2672503/78130742-7fe1-46c3-8bdc-fcd9cefcab2b)

### Expected Behavior

A `opengraph-image` would be generated for each path returned by `generateStaticParams`, same as the `page.tsx` file would.

### How are you deploying your application? (if relevant)

Statically rendered

<sub>[NEXT-1458](https://linear.app/vercel/issue/NEXT-1458/opengraph-images-are-not-statically-generated-for-dynamic-routes)</sub>

**Comment by joeyiny** [comment]
Having this issue as well.

**Comment by altano** [comment]
@huozhi it looks like you fixed this for server rendered dynamic routes in https://github.com/vercel/next.js/issues/48704. Any chance you could take a look if the fix is just as easy for statically generated dynamic routes?

**Comment by chakkun1121** [comment]
You forgot the `generateStaticParams` function in `opengraph-image.tsx`. But when I add `generateStaticParams` function like below, an error has occurred like below.

``` opengraph-image.tsx
// opengraph-image.tsx
...
+ export function generateStaticParams() {
+   return [{ id: "1" }, { id: "2" }];
+ }
``` 

error
```
Error: Page "/[id]/opengraph-image" is missing "generateStaticParams()" so it cannot be used with "output: export" config.
    at /project/sandbox/node_modules/next/dist/build/index.js:909:59
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Span.traceAsyncFn (/project/sandbox/node_modules/next/dist/trace/trace.js:105:20)
    at async Promise.all (index 6)
    at async /project/sandbox/node_modules/next/dist/build/index.js:802:17
    at async Span.traceAsyncFn (/project/sandbox/node_modules/next/dist/trace/trace.js:105:20)
    at async /project/sandbox/node_modules/next/dist/build/index.js:741:124
    at async Span.traceAsyncFn (/project/sandbox/node_modules/next/dist/trace/trace.js:105:20)
    at async build (/project/sandbox/node_modules/next/dist/build/index.js:184:29)
    at async main (/project/sandbox/node_modules/next/dist/bin/next:156:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

(I updated all the libraries.)
(I'm Japanese, so please forgive me if my English is strange.)

**Comment by altano** [comment]
@chakkun1121 yes this is broken. There is no known workaround.

You can’t use export and opengraph images in dynamic routes. It won’t work.

**Comment by Cretezy** [comment]
I'm getting the same behavior as @chakkun1121 in Next 13.5 now too.

**Comment by beeirl** [comment]
Opengraph Image generation doesn't seem to work with `output: "export"` being set in `next.config.js`. Getting a 404 error when trying to access the og image. Using Next 14.0.2. Any ideas for potential workarounds? cc @huozhi

**Comment by loganzartman** [comment]
This doesn't even work in non-dynamic routes for me. Maybe the problem is that Next.js is generating this route `/opengraph-image/[[...__metadata_id__]]/route`, which obviously won't work with `output: export` unless it has `generateStaticParams()`. If there were some way to remove the randomized ID param for the opengraph-image and have this be a non-parameterized route, maybe that would work.

**Comment by altano** [comment]
@loganzartman yes I believe opengraph images ALSO don't work with catch-all routes (whether `output:export`ing or not). Someone else reported that in https://github.com/vercel/next.js/issues/57403 and https://github.com/vercel/next.js/issues/57349.

Opengraph images are pretty broken right now. There are no known workarounds for these issues.

**Comment by steventhanna** [comment]
I was able to get something working by doing the following:

In my page.js file located: `/app/[[...slug]]/page.js`

```jsx
import Logo from '../opengraph-image.png';

export function generateMetadata() {
    return {
        openGraph: {
            images: [Logo.src]
        }
    }
}
```

I also was not able to get it working just by dropping the file in the tree, after much trial and error I had to manually add this piece.

**Comment by cglacet** [comment]
@loganzartman I think you must have done something else to make this work, for me even with `generateStaticParams` I still get nothing rendered statically (at build time):

```bash
├ λ /[locale]/glossary/[slug]/opengraph-image  0 B                0 B
```

While I have the exact same `generateStaticParams` code in the neighbour page but have it properly generated at build time:

```bash
├ ● /[locale]/glossary/[slug]                  7.62 kB         111 kB
```

**Comment by rusakovic** [comment]
@loganzartman  thank you for your solution. Did you work with no `output: export`, but dynamic? To rebuild OG images?

**Comment by altano** [comment]
@loganzartman neat idea. that worked perfectly for my site: I was able to switch to `output: export` and my OG images load instantly now.

It's a shame the built-in OG support is busted but this hack will do. Thanks!

**Comment by suyalcinkaya** [comment]
@loganzartman Thanks for the solution ❤️

**Comment by artginzburg** [comment]
@loganzartman's solution — https://github.com/vercel/next.js/issues/51147#issuecomment-1842197049 — works as expected. I'd just like to add some details:

- Use `export const dynamicParams = false` in the `route.ts(x)` to make the route actually static. This just serves your closest `not-found.tsx` when the user tries to access an unknown slug, so that you don't have to validate params and call `notFound()`.
- You may want to use `[slug]/route.tsx` instead of `[slug]/splash.png/route.tsx`, and append `.png` or any other extension to the ending of each slug in `generateStaticParams()`, so that the final URLs look natural. I just found `/splash-screens/430x932_3.png` vs  `/splash-screens/430x932_3/splash.png` is kind of an obvious stylistic choice.
- The `route.tsx` works just fine on its own, without a specific page. In my case, generating a page at [slug] and having generateStaticParams() in it was not necessary. Instead, I needed to generate a collection of images that would all be added to the metadata of a layout/page, and would then be chosen by the browser based on media queries.

**Comment by tiavina-mika** [comment]
That's strange that there's no mention of this in the official documentation, even more than a year later.

**Comment by samcx** [comment]
> That's strange that there's no mention of this in the official documentation, even more than a year later.

It does!

> - [Dynamic Routes](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) without generateStaticParams()

- https://nextjs.org/docs/app/building-your-application/deploying/static-exports#unsupported-features

```
Route (app)                              Size     First Load JS
┌ ○ /                                    144 B           114 kB
├ ○ /_not-found                          977 B           115 kB
├ ● /[id]                                144 B           114 kB
├   └ /test
└ ● /[id]/opengraph-image                144 B           114 kB
    └ /test/opengraph-image
+ First Load JS shared by all            114 kB
  ├ chunks/28ae4e96-992211943be5776b.js  56.9 kB
  ├ chunks/90-80d7c3169fac80b5.js        55.2 kB
  └ other shared chunks (total)          1.87 kB

○  (Static)  prerendered as static content
●  (SSG)     prerendered as static HTML (uses generateStaticParams)
```

TL;DR, you need to use `generateStaticParams()` in your `opengraph-image.tsx`, or an error will show on `next build`.

**Comment by Cretezy** [comment]
@samcx I just tried thia again last week, it still does not work.

I'll provide an example repo shortly.

**Comment by Cretezy** [comment]
Actually this seems to be working (although can't get it to work on my Cloudflare Pages site). Will report back if I have more details
